src/wasm-type.h

```diff
@@ -602,6 +602,10 @@ struct TypeBuilder {
   // index `i` to be an immediate subtype of the HeapType being built at index
   // `j`. Does nothing for equirecursive types.
   void setSubType(size_t i, size_t j);
+  // In nominal mode, or for nominal types, declare the HeapType being built at
+  // index `i` to be an immediate subtype of the given HeapType. Does nothing
+  // for equirecursive types.
+  void setSubType(size_t i, HeapType super);
```

Comment:
I don't think we need the previous `setSubType` at all. Would you mind trying to replace it? It would also be good to add some tests showing that your use case of setting the supertype to a previously existing type works, perhaps in test/gtest/type-builder.cpp.